src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -988,29 +1034,87 @@ public virtual TokenValidationResult ValidateToken(string token, TokenValidation
                 }
             }
 
-            try
+            TokenValidationResult tokenValidationResult = decryptedJwt != null ? ValidateJWE(outerToken, decryptedJwt, validationParametersCopy) : ValidateJWS(token, validationParametersCopy);
+
+            if (validationParametersCopy.ConfigurationManager != null)
             {
-                if (tokenParts.Length == JwtConstants.JweSegmentCount)
+                if (tokenValidationResult.IsValid)
                 {
-                    var jwtToken = new JsonWebToken(token);
-                    var decryptedJwt = DecryptToken(jwtToken, validationParametersCopy);
-                    var innerToken = ValidateSignature(decryptedJwt, validationParametersCopy);
-                    jwtToken.InnerToken = innerToken;
-                    var innerTokenValidationResult = ValidateTokenPayload(innerToken, validationParametersCopy);
-                    return new TokenValidationResult
-                    {
-                        SecurityToken = jwtToken,
-                        ClaimsIdentity = innerTokenValidationResult.ClaimsIdentity,
-                        IsValid = true,
-                        TokenType = innerTokenValidationResult.TokenType
-                    };
+                    // Set current configuration as LKG if it exists.
+                    if (validationParametersCopy.Configuration != null && validationParametersCopy.Configuration != validationParametersCopy.ConfigurationManager.LastKnownGoodConfiguration)
+                        validationParametersCopy.ConfigurationManager.LastKnownGoodConfiguration = validationParametersCopy.Configuration;
+
+                    return tokenValidationResult;
                 }
-                else
+                else if (tokenValidationResult.Exception.GetType().Equals(typeof(SecurityTokenInvalidSignatureException))
+                   || tokenValidationResult.Exception.GetType().Equals(typeof(SecurityTokenInvalidSigningKeyException))
+                   || tokenValidationResult.Exception.GetType().Equals(typeof(SecurityTokenInvalidIssuerException))
+                   || tokenValidationResult.Exception.GetType().Equals(typeof(SecurityTokenUnableToValidateException))
+                   || tokenValidationResult.Exception.GetType().Equals(typeof(SecurityTokenSignatureKeyNotFoundException)))
```

Comment:
Do we react on 'SecurityTokenInvalidIssuerException' because issuer in metadata might be corrupted?

Reply:
Yes, that was the idea behind falling back to the LKG/requesting refresh for this particular exception.